design_docs/mock-tx-info.md

```diff
@@ -0,0 +1,112 @@
+# Mocking TxInfo
+
+## Context
+
+Some use-cases like testing account contracts may require mocking the contents of `TxInfo` struct returned
+by `get_tx_info()` syscall.
+
+As of writing of this document, `TxInfo` is defined as follows
+
+```cairo
+struct TxInfo {
+    // The version of the transaction. It is fixed (currently, 1) in the OS, and should be
+    // signed by the account contract.
+    // This field allows invalidating old transactions, whenever the meaning of the other
+    // transaction fields is changed (in the OS).
+    version: felt252,
+    // The account contract from which this transaction originates.
+    account_contract_address: ContractAddress,
+    // The max_fee field of the transaction.
+    max_fee: u128,
+    // The signature of the transaction.
+    signature: Span<felt252>,
+    // The hash of the transaction.
+    transaction_hash: felt252,
+    // The identifier of the chain.
+    // This field can be used to prevent replay of testnet transactions on mainnet.
+    chain_id: felt252,
+    // The transaction's nonce.
+    nonce: felt252,
+}
+```
+
+## Goal
+
+Propose a solution for mocking contexts of `TxInfo`.
+
+## Proposed solution
+
+Introduce a new cheatcode `start_mock_tx_info` (working title) with a following signature
+
+```cairo
+fn start_mock_tx_info(contract_address: felt252, tx_info: NewTxInfo) -> Result::<(), felt252> nopanic
+
+struct NewTxInfo {
+    version: Option<felt252>,
+    account_contract_address: Option<felt252>,
+    max_fee: Option<u128>,
+    signature: Array<Option<felt252>>,
+    transaction_hash: Option<felt252>,
+    chain_id: Option<felt252>,
+    nonce: Option<felt252>,
+}
+```
+
+and corresponding `stop_mock_tx_info` cheatcode, that will allow mocking all transactions and calls to the contract
+given by `contract_address`.
+
+`NewTxInfo` accepts `Option`s instead of plain values to allow mocking only parts of `get_tx_info()` response. For
+fields with `None` provided, default values will be used instead of mocking.
+
+Consecutive calls to `start_mock_tx_info` will modify the values of fields where `Some` was provided and leave the
+previous values for fields where `None` was provided.
```

Comment:
A possible problem with this approach: There's no way of cancelling mocks for specific fields. It's only possible to cancel entire `TxInfo`. 

But resolving this would conflict with the ability of "re-mocking" only some field: We cannot support both at the same time.

Reply:
I think lot of those would be resolved by default function argument mechanism in cairo, i think it's planned for the future (don't fact-check me though, im not sure)

Reply:
I wouldn't bet on that, Rust does not support default arguments

Reply:
I've updated the document and I think the new approach provides a good flexibility of mocking even without default arguments.